Move builds to Fedora:40

Fedora:40 is out: https://fedoramagazine.org/announcing-fedora-linux-40/ that means Fedora:39 (what we use in most builds) will be EOL around 2024-11: https://endoflife.date/fedora

Time to start updating the builds. Start by making a list:

- [x] The demo-install build (ty @coryan)
- [x] The m32 build
- [x] The bazel build
	- [x] Handle future_error constructor being removed (https://github.com/llvm/llvm-project/commit/d2cb198f25c82bf77bb113763771590cc79a21a4)
- [x] The Cmake build
	- [x] Fix the clang tidy issues
- [x] The cxx20 build
	- [x] Fix the absl log build error 	
- [x] The cxx14 build
- [ ] #14207
	- [ ] Fix the docfx to handle "operator<<"	
	- [ ] https://github.com/doxygen/doxygen/issues/10859
- [x] The msan build
	- [x] Update the msan LLVM from 17 to 18
- [x] The checkers build

If I was to do this again, I would make a script to make and submit all the PRs to run the builds in GCB. Then go and fix one by one.

For the publish-docs build

This does not generate docs 

```
FAILED: google/cloud/CMakeFiles/cloud-docs /workspace/cmake-out/google/cloud/CMakeFiles/cloud-docs 
cd /workspace/google/cloud && /usr/bin/cmake -E make_directory /workspace/cmake-out/google/cloud && /usr/bin/doxygen /workspace/cmake-out/google/cloud/Doxyfile.cloud-docs
/workspace/google/cloud/doc/error-handling.dox:26: error: unable to resolve reference to 'google::cloud::operator<<(std::ostream&,google::cloud::Status const&)' for \ref command (warning treated as error, aborting now)
T+15.085s [...n_impl.cc.o
ninja: build stopped: subcommand failed.
==> 🕑 publish-docs completed in 239.704 seconds
```

I think I need to follow something like https://github.com/googleapis/google-cloud-cpp/pull/13852 first

--- a/ci/cloudbuild/builds/publish-docs.sh
+++ b/ci/cloudbuild/builds/publish-docs.sh
@@ -48,6 +48,7 @@ doc_args=(
   "-DGOOGLE_CLOUD_CPP_DOXYGEN_CLANG_OPTIONS=-resource-dir=$(clang -print-resource-dir)"
   "-DGOOGLE_CLOUD_CPP_COMMIT_SHA=${COMMIT_SHA}"
   "-DGOOGLE_CLOUD_CPP_ENABLE_CLANG_ABI_COMPAT_17=ON"
+  "-DGOOGLE_CLOUD_CPP_STORAGE_ENABLE_GRPC=ON"
 )
 if command -v /usr/local/bin/sccache >/dev/null 2>&1; then
   doc_args+=(

--- a/ci/cloudbuild/dockerfiles/fedora-latest-publish-docs.Dockerfile
+++ b/ci/cloudbuild/dockerfiles/fedora-latest-publish-docs.Dockerfile
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM fedora:39
+FROM fedora:40
 ARG NCPU=4
 ARG ARCH=amd64
 
@@ -38,13 +38,20 @@ RUN dnf makecache && \
         yaml-cpp-devel
 
 # This is used in the `publish-docs` build
-RUN dnf makecache && dnf install -y libxslt doxygen
+RUN dnf makecache && dnf install -y libxslt
 
 # Sets root's password to the empty string to enable users to get a root shell
 # inside the container with `su -` and no password. Sudo would not work because
 # we run these containers as the invoking user's uid, which does not exist in
 # the container's /etc/passwd file.
 RUN echo 'root:' | chpasswd
+
+# We would prefer to say `dnf install -y doxygen`, but Fedora 40 ships with
+# Doxygen 1.10.0 and we need fixes that were released in Doxygen 1.11.0.
+WORKDIR /var/tmp/doxygen
+RUN curl -fsSL https://github.com/doxygen/doxygen/releases/download/Release_1_11_0/doxygen-1.11.0.linux.bin.tar.gz | \
+    tar -zxf - --strip-components=1 && \
+    make install
 
 WORKDIR /var/tmp/build/
 RUN curl -fsSL https://github.com/open-telemetry/opentelemetry-cpp/archive/v1.15.0.tar.gz | \

--- a/ci/cloudbuild/dockerfiles/fedora-latest-publish-docs.Dockerfile
+++ b/ci/cloudbuild/dockerfiles/fedora-latest-publish-docs.Dockerfile
@@ -47,11 +47,20 @@ RUN dnf makecache && dnf install -y libxslt
 RUN echo 'root:' | chpasswd
 
 # We would prefer to say `dnf install -y doxygen`, but Fedora 40 ships with
-# Doxygen 1.10.0 and we need fixes that were released in Doxygen 1.11.0.
+# Doxygen 1.10.0 and we need fixes that were released in Doxygen 1.11.0. Also,
+# we cannot use the binary, as it was built with clang support, which does not
+# work on Fedora.
+#
+# We follow Doxygen's "compiling from source on UNIX" instructions:
+# https://doxygen.nl/manual/install.html#install_src_unix
 WORKDIR /var/tmp/doxygen
-RUN curl -fsSL https://github.com/doxygen/doxygen/releases/download/Release_1_11_0/doxygen-1.11.0.linux.bin.tar.gz | \
+RUN dnf install -y flex bison
+RUN curl -fsSL https://github.com/doxygen/doxygen/archive/refs/tags/Release_1_11_0.tar.gz | \
     tar -zxf - --strip-components=1 && \
-    make install
+    cmake \
+        -DCMAKE_BUILD_TYPE=Release \
+        -S . -B cmake-out -GNinja && \
+    cmake --build cmake-out --target install
 
 WORKDIR /var/tmp/build/
 RUN curl -fsSL https://github.com/open-telemetry/opentelemetry-cpp/archive/v1.15.0.tar.gz | \

